main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -329,6 +337,32 @@ object Safety {
     case Predicate.Body.Loop(_, _, _) => Set.empty
   }
 
+  /**
+    * Variables that are bound in lattice position without `fix`.
+    */
+  private def forcedLatticeVariables(c0: TypedAst.Constraint): Set[Symbol.VarSym] =
+    c0.body.flatMap(forcedLatticeVariables).toSet
+
+  private def forcedLatticeVariables(p0: Predicate.Body): Set[Symbol.VarSym] = p0 match {
+    case Predicate.Body.Atom(_, Denotation.Latticenal, _, Fixity.Loose, terms, _, _) =>
+      terms.lastOption.map(freeVarsOf).getOrElse(Set.empty)
+    case _ => Set.empty
+  }
+
+  private def checkHeadPredicate(head: TypedAst.Predicate.Head, latVars: Set[Symbol.VarSym]): List[CompilationMessage] = head match {
+    case Head.Atom(_, Denotation.Latticenal, terms, _, loc) =>
+      checkTerms(terms.dropRight(1), latVars, loc)
+    case Head.Atom(_, Denotation.Relational, terms, _, loc) =>
+      checkTerms(terms, latVars, loc)
+  }
+
+  private def checkTerms(terms: List[Expression], latVars: Set[Symbol.VarSym], loc: SourceLocation): List[CompilationMessage] =
+    terms.flatMap(freeVars(_).keys).flatMap { sym =>
```

Comment:
This code is a bit hard to understand. May propose a `checkTerm` function. I think that will clear it up. Reading two flatMap calls is difficult esp. with the `freeVars(_)` thingy.